kubespawner/objects.py

```diff
@@ -566,3 +610,99 @@ def make_ingress(
     )
 
     return endpoint, service, ingress
+
+
+def make_secret(
+    name,
+    env=None,
+    labels=None,
+    annotations=None,
+):
+    """
+    Make a k8s secret specification for mount ssl credentials for running a user notebook.
+
+    Parameters
+    ----------
+    name:
+        Name of the secret. Must be unique within the namespace the object is
+        going to be created in.
+    env:
+        Dictionary of environment variables.
+    labels:
+        Labels to add to the spawned pod.
+    annotations:
+        Annotations to add to the spawned pod.
+
+    """
+
+    if not ssl_enabled(env):
+        return None
+
+    secret = V1Secret()
+    secret.kind = "Secret"
+    secret.api_version = "v1"
+    secret.metadata = V1ObjectMeta()
+    secret.metadata.name = name
+    secret.metadata.annotations = (annotations or {}).copy()
+    secret.metadata.labels = (labels or {}).copy()
+
+    secret.data = {}
+    with open(env['JUPYTERHUB_SSL_KEYFILE'], 'r') as file:
+        encoded = base64.b64encode(file.read().encode("utf-8"))
+        secret.data['key'] = encoded.decode("utf-8")
+    with open(env['JUPYTERHUB_SSL_CERTFILE'], 'r') as file:
+        encoded = base64.b64encode(file.read().encode("utf-8"))
+        secret.data['crt'] = encoded.decode("utf-8")
+    with open(env['JUPYTERHUB_SSL_CLIENT_CA'], 'r') as file:
+        encoded = base64.b64encode(file.read().encode("utf-8"))
+        secret.data["notebooks-ca_trust.crt"] = encoded.decode("utf-8")
+
+    folder = os.path.dirname(env['JUPYTERHUB_SSL_CLIENT_CA'])
+    with open(folder+'/hub-ca_trust.crt', 'r') as file:
+        encoded = base64.b64encode(file.read().encode("utf-8"))
+        secret.data["notebooks-ca_trust.crt"] = secret.data["notebooks-ca_trust.crt"] + encoded.decode("utf-8")
+
+    return secret
+
+
+def make_service(
+    name,
+    port,
+    labels=None,
+    annotations=None,
+):
+    """
+    Make a k8s service specification for using dns to communicate with the notebook.
+
+    Parameters
+    ----------
+    name:
+        Name of the service. Must be unique within the namespace the object is
+        going to be created in.
+    env:
+        Dictionary of environment variables.
+    labels:
+        Labels to add to the spawned pod.
+    annotations:
+        Annotations to add to the spawned pod.
+
+    """
+
+    service = V1Service(
+        kind='Service',
+        metadata=V1ObjectMeta(
+            name=name,
+            annotations=(annotations or {}).copy(),
+            labels=(labels or {}).copy()
+        ),
+        spec=V1ServiceSpec(
+            type='ClusterIP',
+            ports=[V1ServicePort(port=port, target_port=port)],
+            selector={
+                'component': 'singleuser-server',
+                'hub.jupyter.org/username': labels['hub.jupyter.org/username']
```

Comment:
username + component is not unique, since there can be named servers. Maybe this needs `hub.jupyter.org/servername` as well? Or maybe this should just be the pod name instead, which is already unique?

Reply:
So can the user have 2 servers running at once?  If so than the service name is also non-unique currently and will also need to change?

Reply:
From attempting to use Jupyter Hub I can't seem to launch 2 user pods at once.  If that's the case this single service looks like it "should" only have one component it points to.

Reply:
I started my notebook using one spawner config and checked it's labels and had to stop it to start a second one and checked it's labels.

  labels:
    app: jupyterhub
    chart: jupyterhub-0.0.1-set.by.chartpress
    component: singleuser-server
    heritage: jupyterhub
    hub.jupyter.org/network-access-hub: "true"
    hub.jupyter.org/username: jupyter-shane-2estarcher-40xxx-2ecom
    release: jupyterhub

  labels:
    app: jupyterhub
    chart: jupyterhub-0.0.1-set.by.chartpress
    component: singleuser-server
    heritage: jupyterhub
    hub.jupyter.org/network-access-hub: "true"
    hub.jupyter.org/username: jupyter-shane-2estarcher-40xxx-2ecom
    release: jupyterhub